### PR TITLE
Reduxify: Reader Streams

### PR DESCRIPTION
**Pull request by samouri**
**tldr**

We are almost done porting all of Reader state from [Flux to Redux](https://github.com/Automattic/wp-calypso/issues/10984).  This PR aims to port reader Streams.

There has already been some [prior work](https://github.com/Automattic/wp-calypso/pull/19591) to land boilerplate reducer and data-layer code.  Here I present the rest of it. 

**notes on current flux implementation located in: `client/lib/feed-stream`**

---

Its a flux store with many bells and whistles. The API is exposed from two places: `./actions.js` for dispatching (and making api calls) and `./index.js` for accessing a stream from ui code.  It is flexible in that it allows you to request any stream by a key of the format: `streamType[:streamVar1:streamVar2:streamVarN]`.  Two examples would be: `following` or `search:chickens`. Streams know how to:

1. fetch next pages and attach interesting meta depending on the stream type
2. navigate streams via keyboard shortcuts
3. dismiss recommendations via dismissSite api
4. request to fill a gap (and receive updates)
5. poll for updates and then section off new posts so that they aren't released until an update pill is clicked
6. filter out duplicate x-posts before we (I) [broke that it](https://github.com/Automattic/wp-calypso/issues/14722).
7. Interesting behavior in order to also accomodate conversations stream. it adds in comments to `postKeys`

What worked well:

- creating streams and working with them from the UI is very easy. Since the flux store is a one-stop-shop, there is basically one object that encompasses getters, setters, api fetching, modification, state updating, etc. 
-  Flux stores don't go in build

What was hard:

- With half of our state in redux and half in flux, we've been having the awkwardness of using two paradigms at once.
- so much functionality in a single place means it was approaching a :spaghetti: state.  
- it was pretty late when we started extracting useful utility functions like `post-keys.js`, we can definitely do better in a v2.

**proposed state**

In order to satisfy the requirements for `InfiniteList`, we need to provide this state-related props: `items: Array<StreamItems>`, `lastPage: boolean`, `fetchNextPage: Function`, `fetchingNextPage: boolean`.  In the redux world we can see these all being selectors. 

We can move all of the getters on the current store to selectors as well.
We can move all the data-fetching to the data-layer. 
All data storage cleanly maps to reducers

Types/State: 

```typescript
type StreamItem = PostKey | Gap | RecommendationsKey | CombinedCardKey;

type PostKey = {
    type: 'postKey'
    feedId?: number,
    blogId?: number,
    postId: number,
}

type ConversationsPostKey = {
    ...PostKey,
    type: 'conversationPostKey',
    comments: [ id markers for maxId of comment to show ]
}

type Gap = {
    type: 'gap',
    from: date, // maybe PageHandles or PostKeys in a future iteration
    to: date,
}

type RecommendationsKey = {
    type: 'recommendations',
    postKeys: Array<PostKey>
}

type CombinedCardKey = {
    type: 'combined-card',
    postKeys: Array<PostKey>
}

type StreamKey = string;

state.reader.streams = {
    // object containing a map of StreamKey: StreamData
    items: { 
        [ streamKey ]: {
            selected: StreamItem | null, // only relevant to keyboard navigation 
            items: Array<StreamItems>, // list of all known 
            pendingItems: { 
                 items: Array<StreamItem>, // taken advantage of by the update-pill so that we can know about items without having to render them.
                 lastUpdated: Date // the time the stream was last updated so that pendingItems knows what to pay attention to
            }
            lastPage: boolean, // whether the stream is at the end or not
            isRequesting: boolean, // whether stream is currently requesting next page
            pageHandle: object, // the query parameters to use to fetch the next page of data
            
        }
    },
    curentStream: StreamKey, // only relevant to keyboard navigation from within full-post page  
}
```

### CODE REVIEW

**Comment by blowery** [comment]
I think the post keys currently also carry either a date, or a tagged on date, or a liked on date... That's to allow us to request the next page without having to pull the post from the post store (and sometimes the post store may not _have_ it, if the post was reloaded in a different stream).

**Comment by blowery** [comment]
`state.reader.streams.currentStream` could live over in the UI reducers too.

**Comment by blowery** [comment]
overall I like it. Seems like a good plan.

**Comment by blowery** [comment]
There's also some current crazy stuff with recs where we fire train-tracks events when the stream items come in. Maybe we do that in the data layer?

**Comment by samouri** [comment]
> I think the post keys currently also carry either a date, or a tagged on date, or a liked on date... That's to allow us to request the next page without having to pull the post from the post store (and sometimes the post store may not have it, if the post was reloaded in a different stream).

I'm experimenting with storing a `pageHandle` in redux to solve that problem

**Comment by samouri** [comment]
I've reached a milestone! All streams correctly load and page. 

Still to do:
- [ ] handle updates and pills for date streams
- [ ] stream transformations (recs and combinations mainly)
- [ ]

File: client/reader/stream/utils.js [hunk 1]
```diff
@@ -68,6 +68,7 @@ export function combine( postKey1, postKey2 ) {
 	return combined;
 }
 
+//@TODO: operate on things with dates...
```
Comment (samouri) [inline client/reader/stream/utils.js]: no idea what i was talking about here...maybe i already fixed it since CCes work

**Comment by samouri** [comment]
This is a really big PR that would be incredibly difficult to split up.  Streams are the hairiest part of all of Reader and are also deeply depended on by almost every UI. To split this up would mean partially relying on redux streams and partially relying upon flux streams which I believe would be messier than one difficult PR with a lot of manual review.

To that end, here is a list of everything to test:
- Search Streams: smoke test by performing a search. scrolling through. keyboard navigation. Switching sort order and verifying the timestamps of the date sorted results.
-

**Comment by samouri** [comment]
Two additional nice to haves that I'm punting for future PRs:
1. We should be able to get rid of the custom infinite-stream implementations inside of `/search` recommendations and in `following-manage` in favor of reusing this implementation.
2. We should be able to filter out duplicate x-posts now

**Comment by bluefuton** [comment]
Tremendous. Thanks for taking this on @samouri 💯 

I gave it a good test drive and stream paging feels snappier! I only found one issue and this happened on list and tag streams only:

screen shot 2018-05-03 at 5 47 57 pm

Specifically, it happened for me in:

http://calypso.localhost:3000/read/list/bluefuton/2016-02
http://calypso.localhost:3000/tag/bananaaaa

File: client/reader/controller.js [hunk 2]
```diff
@@ -176,24 +171,20 @@ const exported = {
...
-		const basePath = '/read/feeds/:feed_id',
-			fullAnalyticsPageTitle = analyticsPageTitle + ' > Feed > ' + context.params.feed_id,
-			feedStore = feedStreamFactory( 'feed:' + context.params.feed_id ),
-			mcKey = 'blog';
-
-		ensureStoreLoading( feedStore, context );
+		const basePath = '/read/feeds/:feed_id';
+		const fullAnalyticsPageTitle = analyticsPageTitle + ' > Feed > ' + context.params.feed_id;
+		const mcKey = 'blog';
+		const feedId = context.params.feed_id;
```
Comment (blowery) [inline client/reader/controller.js]: moving this higher would make the fullAnalyticsPageTitle simpler

File: client/reader/like-helper.jsx [hunk 3]
```diff
@@ -1,22 +1,18 @@
...
 		}
+	} else if ( post && post.site_ID && ! post.is_external ) {
+		showLikes = true;
+	}
 
-		return showLikes;
-	},
-};
-
-export default exported;
```
Comment (blowery) [inline client/reader/like-helper.jsx]: yaaaaaaay

File: client/reader/site-stream/featured.jsx [hunk 4]
```diff
@@ -94,15 +103,15 @@ class FeedFeatured extends React.PureComponent {
 	}
 }
 
-function mapStateToProps( state, ownProps ) {
-	const { postsStore } = ownProps;
-	const postKeys = postsStore.get();
+function mapStateToProps( state, { streamKey } ) {
+	const stream = getStream( state, streamKey );
+	const postKeys = stream.items;
 	const posts = getPostsByKeys( state, postKeys );
```
Comment (blowery) [inline client/reader/site-stream/featured.jsx]: not really this PR, but are these creating new arrays on every mapping? Should probably be in a memoized selector instead.

File: client/reader/stream/empty-search-recommended-post.jsx [hunk 5]
```diff
@@ -26,10 +26,11 @@ export default function EmptySearchRecommendedPost( { post } ) {
 		recordAction( 'search_page_rec_site_click' );
 	}
 
-	const site = { title: post.site_name };
+	const site = { title: post && post.site_name };
 
+	/* eslint-disable  wpcalypso/jsx-classname-namespace */
 	return (
-		<div className="search-stream__recommendation-list-item" key={ post.global_ID }>
+		<div className="search-stream__recommendation-list-item" key={ post && post.global_ID }>
```
Comment (blowery) [inline client/reader/stream/empty-search-recommended-post.jsx]: if there's no post, there's no key... 

should the parent be assigning the key?

File: client/reader/stream/index.jsx [hunk 6]
```diff
@@ -325,60 +255,74 @@ class ReaderStream extends React.Component {
...
 
 	showUpdates = () => {
+		const { streamKey } = this.props;
 		this.props.onUpdatesShown();
-		showUpdates( this.props.postsStore.id );
-		if ( this.props.recommendationsStore ) {
-			shufflePosts( this.props.recommendationsStore.id );
-		}
+		this.props.showUpdates( { streamKey } );
+		// @todo: do we need to shuffle?
```
Comment (blowery) [inline client/reader/stream/index.jsx]: the shuffle was to prevent us from showing the same recs in the same relative order after a page "refresh"... unsure if we really need it.

File: client/reader/stream/index.jsx [hunk 7]
```diff
@@ -470,9 +409,13 @@ class ReaderStream extends React.Component {
 			);
 			showingStream = true;
 		}
+		const indexOfColon = streamKey.indexOf( ':' );
```
Comment (blowery) [inline client/reader/stream/index.jsx]: maybe make a helper that returns a pair of stream type and stream id?

File: client/reader/stream/index.jsx [hunk 8]
```diff
@@ -470,9 +409,13 @@ class ReaderStream extends React.Component {
...
+		const streamType = indexOfColon === -1 ? streamKey : streamKey.substring( 0, indexOfColon );
+
 		const TopLevel = this.props.isMain ? ReaderMain : 'div';
 		return (
 			<TopLevel className={ classnames( 'following', this.props.className ) }>
+				{ ! streamType === 'search' && <Interval onTick={ this.poll } period={ EVERY_MINUTE } /> }
```
Comment (blowery) [inline client/reader/stream/index.jsx]: `streamType !== 'search'` would be simpler i think

File: client/reader/stream/index.jsx [hunk 9]
```diff
@@ -481,25 +424,50 @@ class ReaderStream extends React.Component {
 					) }
 
 				<UpdateNotice
-					count={ this.state.updateCount }
+					count={ updateCount }
```
Comment (blowery) [inline client/reader/stream/index.jsx]: can we connect up updateNotice and pass it the stream key?

File: client/state/data-layer/wpcom/read/streams/index.js [hunk 10]
```diff
@@ -111,39 +172,85 @@ const streamApis = {
...
-	const { streamKey, query } = action.payload;
-	return receivePage( { streamKey, query, posts } );
+export function handlePage( action, data ) {
+	const { posts, date_range, meta, next_page } = data;
+	const { streamKey, query, isPoll, gap, streamType } = action.payload;
+	const { dateProperty } = streamApis[ streamType ];
+	let pageHandle = {};
+
+	if ( meta && meta.next_page ) {
+		// sites give pange handles nested within the meta key
```
Comment (blowery) [inline client/state/data-layer/wpcom/read/streams/index.js]: s/pange/page/

File: client/state/reader/streams/actions.js [hunk 11]
```diff
@@ -26,41 +26,46 @@ import {
...
+			streamItems,
 			pageHandle,
+			gap,
 		},
 	};
 }
 
-export function showUpdates( { streamKey } ) {
-	return {
+export const showUpdates = ( { streamKey } ) => ( dispatch, getState ) => {
```
Comment (blowery) [inline client/state/reader/streams/actions.js]: why does this need to be a thunk?

Comment (blowery) [reply client/state/reader/streams/actions.js]: ahhh, because the pending items are outside the scope of the current items reducer

File: client/state/reader/streams/reducer.js [hunk 12]
```diff
@@ -16,50 +17,104 @@ import {
...
 export const items = ( state = [], action ) => {
 	switch ( action.type ) {
 		case READER_STREAMS_PAGE_RECEIVE:
-			const { posts } = action.payload;
-			const postKeys = posts.map( keyForPost );
+			const { streamItems, gap } = action.payload;
 
-			const newState = uniqBy( [ ...state, ...postKeys ], keyToString );
+			let nextState;
+			if ( !! gap ) {
```
Comment (blowery) [inline client/state/reader/streams/reducer.js]: why the `!!`?

File: client/state/reader/streams/reducer.js [hunk 13]
```diff
@@ -16,50 +17,104 @@ import {
...
+				}
+
+				nextState = [ ...beforeGap, ...streamItems, ...nextGap, ...afterGap ];
+			} else {
+				nextState = [ ...state, ...streamItems ];
 			}
+
+			nextState = uniqWith( nextState, keysAreEqual );
+
+			if ( nextState.length === state.length ) {
```
Comment (blowery) [inline client/state/reader/streams/reducer.js]: this doesn't seem safe...

Comment (blowery) [reply client/state/reader/streams/reducer.js]: thought maybe it is. would it catch the case where a gap is replaced with one post?

File: client/reader/stream/index.jsx (hunk 6 above)
Comment (samouri) [reply client/reader/stream/index.jsx]: depends on what you mean by page refresh. If the browser is actually refreshed then they will see brand new recs due to a new seed.

File: client/reader/stream/index.jsx (hunk 7 above)
Comment (samouri) [reply client/reader/stream/index.jsx]: i actually have a bunch of unpushed changes right now that extract this function into one called `getStreamType`

File: client/reader/stream/index.jsx (hunk 8 above)
Comment (samouri) [reply client/reader/stream/index.jsx]: agreed! thanks i'll update

File: client/reader/stream/index.jsx (hunk 9 above)
Comment (samouri) [reply client/reader/stream/index.jsx]: i'll look into it! seems like that should work to me

File: client/state/reader/streams/actions.js (hunk 11 above)
Comment (samouri) [reply client/state/reader/streams/actions.js]: yea....its the problem with `combineReducers` IMO.  There are a lot of alternatives equally meh solutions I thought of.  One of them is for the react component to pass the action creator the `pendingItems`.  Let me know if you have any ideas

File: client/state/reader/streams/reducer.js (hunk 12 above)
Comment (samouri) [reply client/state/reader/streams/reducer.js]: because its so exciting!!!!! I'll remove it :)

File: client/state/reader/streams/reducer.js (hunk 13 above)
Comment (samouri) [reply client/state/reader/streams/reducer.js]: Thats a good point, I can make this optimization safer and always return `nextState`

File: client/reader/site-stream/featured.jsx (hunk 4 above)
Comment (samouri) [reply client/reader/site-stream/featured.jsx]: `getPostsByKeys` is memoized :)

File: client/reader/stream/index.jsx [hunk 14]
```diff
@@ -89,71 +86,28 @@ class ReaderStream extends React.Component {
 		showMobileBackToSidebar: true,
 		isDiscoverStream: false,
 		shouldCombineCards: true,
-		transformStreamItems: identity,
 		isMain: true,
 		useCompactCards: false,
 		intro: null,
 		forcePlaceholders: false,
 	};
+	page = 0;
```
[the remark this answers is not in the document]
Comment (samouri) [reply client/reader/stream/index.jsx]: nope! I only put things in state if its modification would then require a rerender.  `page` is just for analytics and changing it shouldn't cause a rerender

File: client/blocks/reader-full-post/index.jsx [hunk 15]
```diff
@@ -491,13 +483,15 @@ export class FullPostView extends React.Component {
 export default connect(
 	( state, ownProps ) => {
 		const { feedId, blogId, postId } = ownProps;
-		const post = getPostByKey( state, { feedId, blogId, postId } ) || { _state: 'pending' };
+		const postKey = pickBy( { feedId: +feedId, blogId: +blogId, postId: +postId } );
```
[the remark this answers is not in the document]
Comment (samouri) [reply client/blocks/reader-full-post/index.jsx]: I wish it read better. `pickBy` without a second arg essentially means `omitFalsyValues`

File: client/reader/controller.js (hunk 2 above)
Comment (samouri) [reply client/reader/controller.js]: will do

File: client/reader/stream/index.jsx [hunk 16]
```diff
@@ -214,48 +163,31 @@ class ReaderStream extends React.Component {
...
-			nextProps.postsStore.on( 'change', this.updateState );
-			nextProps.recommendationsStore &&
-				nextProps.recommendationsStore.on( 'change', this.updateState );
+		const { streamKey } = nextProps;
+		if ( streamKey !== this.props.streamKey ) {
 			this.props.resetCardExpansions();
-
-			this.updateState( nextProps );
-			this._list && this._list.reset();
+			this.props.viewStream( { streamKey } );
```
[the remark this answers is not in the document]
Comment (samouri) [reply client/reader/stream/index.jsx]: damn, you just made me realize that `page` doesn't belong anywhere in React per se.  It isn't tied to rerenders, or the mounting/dismounting of components.  It shouldn't be reset when `streamKey` is changed. There needs to be a separate page held on to by each streamKey that is incremented (and remembered).

I'll move this logic to the `analyticsAlgoMap` that I've already created in the data-layer, which is a better fit anyway.

[the remark this answers is not in the document]
Comment (samouri) [reply client/reader/stream/index.jsx]: I started refactoring all of the analytics that occur in `Streams` to isolate where they live instead of existing partially in a few places (data-layer, stream component, and controller).  In particular it would be ideal if the config for analytics per stream lived in just one place.

As I was making the sweeping changes I realized that it was going to be much too large to review on top of this already massive PR.  I've started the work in a different branch, and for the purposes of this PR I think we can be best served by maintaining page counts in a simple map at the module level of the `Stream` react component.

**Comment by samouri** [comment]
ps: i'm currently working on refactoring some of the logic with accepting updates to hopefully simplify it, so until the EOD today expect a bug or two

**Comment by samouri** [comment]
ps: i'm currently working on refactoring some of the logic with accepting updates to hopefully simplify it, so until the EOD today expect a bug or two

**Comment by samouri** [comment]
ps: i'm currently working on refactoring some of the logic with accepting updates to hopefully simplify it, so until the EOD today expect a bug or two

**Comment by samouri** [comment]
ps: i'm currently working on refactoring some of the logic with accepting updates to hopefully simplify it, so until the EOD today expect a bug or two

**Comment by blowery** [comment]
Seems like the number of placeholders when the stream is empty has dropped to two, from ... 4? I think more looks better. Maybe just use 7 all the time?

**Comment by samouri** [comment]
> Seems like the number of placeholders when the stream is empty has dropped to two, from ... 4? I think more looks better. Maybe just use 7 all the time?

I had modified the fetching counts to always be 6 and didn't update the placeholders to match it. After some discussion we decided to match old behavior even if just for the sake of maintaining consistent metrics in this PR. To that end, the new (and old) behavior is: 4 posts for initial fetch, 7 for subsequent fetches.
